File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start)
{
    if (start < 1)
    {
        return ERROR_VALUE;
    }

    int count = 0;

    int num = start;
    for (;;)
    {
        if (num == 1)
        {
            return count;
        }

        count++;
        if (num % 2 == 0)
        {
            num /= 2;
        }
        else
        {
            num = (3 * num) + 1;
        }
    }
}

File: zig/collatz-conjecture/collatz_conjecture.zig
pub const ComputationError = error{IllegalArgument};

pub fn steps(number: usize) anyerror!usize {
    if (number == 0) return ComputationError.IllegalArgument;
    
    var count: usize = 0;
    var n = number;
    while (n > 1) {
        if (n % 2 == 0) {
            n /= 2;
        } else {
            n = 3 * n + 1;
        }
        count += 1;
    }
    return count;
}
